### DISCUSSION TITLE
Guard.NET

### DISCUSSION THREAD

**Discussion by gverstraete** [arcus-azure/arcus#362] [General]
Hi,

We are experiencing some memory leaks because of Guard.NET in my project. It takes up to 100MB.

It also seems that the GitHub repo is not very active anymore: https://github.com/george-pancescu/Guard.

Question here is: should we maintain to use Guard in Arcus or should we move away from it?

We decided to implement Guard on our own.

**Comment by fgheysels** [comment]
Thanks for bring this to our attention.

The fact that it hasn't received any recent updates is maybe related to the fact that this library doesn't have that much functionality.  Next to that, Guard.NET supports .NET standard, so it supports a whole lot of .NET versions.
Anyway, my opinion here is also that this library adds yet another dependency / package to Arcus, and if we want to keep Arcus as lightweight as possible, I wonder why we're not just use native code to guard for pre-conditions / null checks ?

With more recent versions of .NET, we can simplify null checks as well with built-in functionality, like this:
```
ArgumentNullException.ThrowIfNull(parameter);
```

And in C# 11, we can even just write it like this:

```
private void DoSomething( object someParameter!! )
{
}
```
See [here](https://www.stevefenton.co.uk/blog/2022/03/parameter-null-checking-in-c/#what-is-parameter-null-checking)

So, my take would be to slowly move away from using Guard.NET in Arcus :)  What's your opinion @stijnmoreels  ?

**Reply by fgheysels** [reply]
I'm also wondering: are we absolutely sure that the mem-leak was to be found in Guard.NET ?
I would assume that, if Guard.NET has a mem-leak, that this would've been reported in the Guard.NET repo already ?  
If not, and if you can pinpoint where the mem-leak exists, maybe open an issue in the Guard.NET repo as well  @gverstraete ?

**Reply by gverstraete** [reply]
I wasn't involved in the deep analysis. But the feedback i got was this:

> is a memory-hungry framework.
Instead of returning immediately when the object under validation is ok, it allocates a string with the error, Instantiates an Exception, and does nothing with it.
We allocate memory with strings and Exception Objects for Nothing across all solutions.

We have some very big applications & because of this it was taking almost 100 MB of allocations in some cases.

**Reply by fgheysels** [reply]
Discussed this offline with @stijnmoreels and we agreed to phase out the use of Guard.NET in Arcus.
